### DISCUSSION TITLE
How do I deploy Bun.js applications

### DISCUSSION THREAD

**Discussion by nhuxhr** [oven-sh/bun#559] [Q&A]
I wanna start using Bun.js and drop Node.js depending of the application I wanna develop tho.

But I'm a little bit confused on how I'm gonna deploy Bun.js applications, I'm not sure if I can deploy Bun.js applications on Heroku, Vercel (for Next.js apps), Netlify (for React.js apps), or maybe my own shared hosting that supports Node.js apps.

I'm using a Windows PC tho but I have decide to install Linux for Bun.js sake, My PC is capable of Linux tho, I just don't see any use for it yet but here comes Bun.js!

Looking forward to hearing from you.

Thank you :)

**Comment by Bellisario** [comment]
You can't deploy directly a Bun.js application to (for ex.) Heroku, because these services use Node.js while you're using Bun.js.

Theorically, they're all javascript, but some APIs like the fetch API (from Bun.js homepage: `JavaScript runtime with Web APIs like fetch`) are not Node.js APIs always available ([reading here](https://blog.logrocket.com/fetch-api-node-js/) seems to be integrated only on 17.5 and above, which is doesn't come to an LTS version, yet) and, another common example can be the top level "await", which is not supported by Node.js, too.

Bun.js it's on Beta and for now shouldn't be used for production but just for "fun", because something could not work properly...

Anyway, there are two things you can do to really deploy a Bun.js application:
- Check it with Node.js (running with it) before deploy (but since you need to always use both Bun.js and Node.js and if something is broken you need to find the way you can make it working on Node.js, you're building a Node.js application using Bun.js and that's not very good);
- Don't deploy to these "limited" servers and instead opt for a real (or in the cloud) machine, install Bun.js to it and then run your application (this is the real solution but it's expensive, since you need a total machine and there is no free cloud provider that I know).

A good thing would be if these (free and not) would choose to offer also Bun.js but for now (also because there is only the Beta) no one does. I expect to see something comparable to [Deno Deploy](https://deno.com/deploy) (but for Bun.js) in the future, but not in a recent time...

**Comment by nhuxhr** [comment]
Alright! Thanks for your response.

Can't wait to start using Bun.js for live projects tho!

I love this project, it looks promising tho 😍

**Comment by AlejandroAkbal** [comment]
Hopefully someone creates a build pack that can be used in Heroku and similar services

**Reply by rubys** [reply]
Not a build pack, but you can deploy Bun applications on fly.io with a Dockerfile.  We will even create the Dockerfile for you.  See: https://fly.io/blog/flydotio-heart-bun/

Note that the Dockerfiles produced can be run not just on fly.io, but anywhere including your own machine.  You don't even have to use `fly launch` to build them, you can use https://github.com/fly-apps/dockerfile-node#usage

**Reply by schneems** [reply]
I work for Heroku. I found the install instructions here https://bun.sh/docs/installation. There are several ways to get this on Heroku. You could create a buildpack and put that bash invocation into the `bin/compile` script https://devcenter.heroku.com/articles/buildpack-api. Though you might need to move the resulting binary around, only files inside the application directory are installed.

Alternatively is it possible to apt get install this? We have an "apt buildpack" https://elements.heroku.com/buildpacks/heroku/heroku-buildpack-apt. 

It looks like below there's a method to install below as well, though I'm presuming that requires `npm` which it might not make sense to stuff both node and bun into a dyno.

**Comment by Dev-Siri** [comment]
Is there a docker image for Bun that can be used to deploy it to something like render?

**Reply by nhuxhr** [reply]
[https://github.com/oven-sh/bun/discussions/1040#discussioncomment-3360549](https://github.com/oven-sh/bun/discussions/1040#discussioncomment-3360549)

**Comment by janwilmake** [comment]
If you use a monorepo you can do something like this:

```
{
  "version": "0.0.0",
  "private": true,
  "workspaces": [
    "packages/scrapingbee/*",
    "packages/helpers/*"
  ],
  "dependencies": {
    "bun": "0.7.0",
    "pm2": "^5.3.0"
  },
  "scripts": {
    "start": "cd packages/server && bun run start"
  }
}
```

This way, you can run bun anywhere including Vercel, Heroku, etc. As far as I'm aware, it's working fine.

**Reply by PatrickMatthiesen** [reply]
This worked surprisingly well. I don't need to install the bun separately on the server. I don't need to use npm install either, as I can have Bun do it in my CI step.
Thanks a lot

**Reply by iamvikshan** [reply]
well, that worked

**Comment by mgan59** [comment]
For heroku support, won't be able to use a buildpack.  Think the approach is to build a container dyno and mount the bun docker container. Similar to how you would deploy on-premise n8n or retool.

**Comment by elycheikhsmail** [comment]
you can use domcloud.co

**Comment by yeukfei02** [comment]
which cloud can use to deploy nodejs with bun?
lets say I use bun workspace.

thanks all.

**Reply by elycheikhsmail** [reply]
domcloud.io

**Comment by naoyanickf** [comment]
I did deployed simple bun app on heroku.

render made docker sample. here
https://github.com/render-examples/bun-docker

I used that. just add heroku.yml

**Comment by olehmisar** [comment]
If you are using yarn, `yarn add bun` will not work. So you can instead put `"start": "npx -y bun src/index.ts"` in your `scripts` in package.json.

**Comment by olawalejuwonm** [comment]
Is there an update on this?

**Reply by bluefoxy009** [reply]
Are the posted solutions above unable to suit your use case? If so, what specifically are you trying to accomplish/what platform are you using that isn't working?

**Comment by jeevaphysicist** [comment]
We can deploy our Bun.js server on the Render hosting platform. This blog helps those who need to deploy a Bun.js app like a Node.js app: https://bun.sh/guides/ecosystem/render.

**Comment by jakeg** [comment]
With Heroku you can use the Bun buildpack I made: https://github.com/jakeg/heroku-buildpack-bun

I use it myself in production, and have been since April 2023.
